File: analysis/import/CSVImporter/src/test/java/de/maibornwolff/codecharta/importer/csv/CSVProjectAdapterTest.java
package de.maibornwolff.codecharta.importer.csv;

import de.maibornwolff.codecharta.model.Node;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.hasSize;

public class CSVProjectAdapterTest {
    private final CSVProjectAdapter project = new CSVProjectAdapter("test", '\\', ',');

    private static InputStream toInputStream(String content) {
        return new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
    }

    @Test
    public void should_ignore_row_if_no_path_column_present() throws UnsupportedEncodingException {
        // when
        project.addProjectFromCsv(toInputStream("head,path\nnoValidContent\n"));

        // then
        assertThat(project.getRootNode().getChildren(), hasSize(0));
    }

    @Test
    public void should_read_node_name_from_specified_path_column() throws UnsupportedEncodingException {
        String name = "someName";
        // when
        project.addProjectFromCsv(toInputStream("someContent,,path\nprojectName,blubb2," + name));

        // then
        List<Node> rootNode = project.getRootNode().getChildren();
        assertThat(rootNode.size(), is(1));
        assertThat(rootNode.iterator().next().getName(), is(name));
    }

    @Test
    public void should_read_node_with_name_only_once() throws UnsupportedEncodingException {
        String name = "someName";
        // when
        project.addProjectFromCsv(toInputStream("someContent\n" + name));
        project.addProjectFromCsv(toInputStream("someContent\n" + name));

        // then
        assertThat(project.getRootNode().getChildren().size(), is(1));
    }

    @Test
    public void should_create_nodes_for_directories() throws UnsupportedEncodingException {
        // given
        String directoryName = "someNodeName";

        // when
        project.addProjectFromCsv(toInputStream("someContent\n" + directoryName + "\\someFile"));

        // then
        assertThat(project.getRootNode().getChildren().size(), is(1));
        Node node = project.getRootNode().getChildren().iterator().next();
        assertThat(node.getName(), is(directoryName));
        assertThat(node.getChildren().size(), is(1));
    }

    @Test
    public void should_read_node_attributes_if_metric_values() throws UnsupportedEncodingException {
        // given
        String attribName = "attname";
        String attribVal = "\"0,1\"";
        float attValFloat = 0.1F;

        // when
        project.addProjectFromCsv(toInputStream("head1,path,head3,head4," + attribName + "\nprojectName,\"9900,01\",\"blubb\",1.0," + attribVal + "\n"));

        // then
        Map<String, Object> nodeAttributes = project.getRootNode().getChildren().iterator().next().getAttributes();
        assertThat(nodeAttributes.size(), is(3));
        assertThat(nodeAttributes.get(attribName), is(attValFloat));
    }
}

File: analysis/import/CSVImporter/src/test/kotlin/de/maibornwolff/codecharta/importer/csv/CSVProjectAdapterTest.kt
package de.maibornwolff.codecharta.importer.csv

import de.maibornwolff.codecharta.translator.MetricNameTranslator
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.nio.charset.StandardCharsets

class FFCSVProjectAdapterTest : Spek({
    fun toInputStream(content: String): InputStream {
        return ByteArrayInputStream(content.toByteArray(StandardCharsets.UTF_8))
    }

    describe("a CSVProjectAdapter") {
        val project = CSVProjectAdapter("test", '\\', ',')

        on("adding invalid csv") {

            val invalidContent = "head,path\nnoValidContent\n"
            project.addProjectFromCsv(toInputStream(invalidContent))

            it("should be ignored") {
                assertThat(project.rootNode.children, hasSize(0))
            }
        }

        on("adding valid csv") {
            val name = "someName"
            project.addProjectFromCsv(toInputStream("someContent,,path\nprojectName,blubb2,$name"))

            it("should have node with same name") {
                assertThat(project.rootNode.children.map { it.name }, hasItem(name))
            }
        }

        on("adding same line twice") {
            val name = "someNameOrOther"
            project.addProjectFromCsv(toInputStream("someContent\n" + name))
            project.addProjectFromCsv(toInputStream("someContent\n" + name))

            it("should add only first line") {
                assertThat(project.rootNode.children.filter { it.name == name }.size, `is`(1))
            }
        }
    }

    describe("a CSVProjectAdapter") {
        val project = CSVProjectAdapter("test", '\\', ',')

        on("adding line with metric values") {
            val attribName = "attname"
            val attribVal = "\"0,1\""
            val attValFloat = 0.1f

            project.addProjectFromCsv(toInputStream("head1,path,head3,head4,$attribName\nprojectName,\"9900,01\",\"blubb\",1.0,$attribVal\n"))

            it("should add attributes to node") {
                val nodeAttributes = project.rootNode.children.iterator().next().attributes
                assertThat(nodeAttributes.size, `is`(3))
                assertThat<Any>(nodeAttributes[attribName], `is`<Any>(attValFloat))
            }
        }

    }

    describe("a CSVProjectAdapter") {
        val project = CSVProjectAdapter("test", '\\', ',')

        on("adding file with subdirectory") {
            val directoryName = "someNodeName"
            project.addProjectFromCsv(toInputStream("someContent\n$directoryName\\someFile"))

            it("should create node for subdirectory") {
                assertThat(project.rootNode.children.size, `is`(1))
                val node = project.rootNode.children.iterator().next()
                assertThat(node.name, `is`(directoryName))
                assertThat(node.children.size, `is`(1))
            }
        }
    }

    describe("CSVProjectAdapter for Sourcemonitor") {
        val project = CSVProjectAdapter("test", '\\', ',',
                MetricNameTranslator(mapOf(Pair("File Name", "path"))))

        on("reading csv lines from Sourcemonitor") {
            project.addProjectFromCsv(this.javaClass.classLoader.getResourceAsStream("sourcemonitor.csv"))

            it("has more than one node") {
                assertThat(project.rootNode.nodes.size, greaterThan(1))
            }
        }
    }

    describe("CSVProjectAdapter for Understand") {
        val project = CSVProjectAdapter("test", '\\', ',',
                MetricNameTranslator(mapOf(Pair("File", "path"))),
                { it[0] == "File" }
        )


        on("reading csv lines from Understand") {
            project.addProjectFromCsv(this.javaClass.classLoader.getResourceAsStream("understand.csv"))

            it("has more than one node") {
                assertThat(project.rootNode.nodes.size, greaterThan(1))
            }
        }
    }
})